src/Controller/Controller.js

```diff
@@ -0,0 +1,35 @@
+import Car from '../Model/Car.js';
+import { getInput, printCarPositions, printWinners } from '../View/View.js';
+import {
+  validateCarNames,
+  validateCars,
+  validateRounds,
+} from '../Validation/Validation.js';
+import getWinners from '../Util/getWinners.js';
+import runRounds from '../Util/runRounds.js';
+import SYSTEM_MESSAGES from '../Model/SystemMessages.js';
+
+// Controller에서 전체 로직을 처리
+
+class Controller {
+  async run() {
+    const carNames = await getInput(SYSTEM_MESSAGES.ASK_CARS_NAME);
+    const validatedCarNames = validateCarNames(carNames);
+
+    const roundsInput = await getInput(SYSTEM_MESSAGES.ASK_ROUNDS);
+    const validatedRounds = validateRounds(roundsInput);
+
+    /** @type {Car[]} */
+    const cars = validatedCarNames.map((carName) => new Car(carName));
+    const validatedCars = validateCars(cars);
+
+    runRounds(validatedCars, validatedRounds, printCarPositions);
+
+    const maxPosition = Math.max(...validatedCars.map((car) => car.position));
+    const winners = getWinners(cars, maxPosition);
+
+    printWinners(winners);
+  }
+}
+
+export default Controller;
```

Comment:
2주차 과제가 끝나고 다른 분들 코드도 참고하면서 이것저것 알아보면서 느낀 점인데 입력값을 비동기적으로 병렬 입력받는 방법은 어떨까요?  추가적으로 로직들을 하나의 메서드로 묶는다면 좀 더 유지보수하기 쉬운 코드가 될 것 같아요! 아래는 제가 생각해본 리팩터링 코드입니다!

```suggestion
class Controller {
  async run() {
    const [carNames, roundsInput] = await this.getInputs();
    const cars = this.initializeCars(carNames);
    const validatedRounds = this.validateRounds(roundsInput);

    this.startRace(cars, validatedRounds);
    this.printResults(cars);
  }

  async getInputs() {
    return Promise.all([
      getInput(SYSTEM_MESSAGES.ASK_CARS_NAME),
      getInput(SYSTEM_MESSAGES.ASK_ROUNDS),
    ]);
  }

  initializeCars(carNames) {
    const validatedCarNames = validateCarNames(carNames);
    return validateCars(validatedCarNames.map((carName) => new Car(carName)));
  }

  startRace(cars, rounds) {
    runRounds(cars, rounds, printCarPositions);
  }

  printResults(cars) {
    const maxPosition = Math.max(...cars.map((car) => car.position));
    const winners = getWinners(cars, maxPosition);
    printWinners(winners);
  }
}
```

Reply:
오... 이방식이 상당히 깔끔해 보이네요. side effect도 적을것 같고요... 다만, 다음에는 서비스를 만들고 의존성 주입으로 코드를 짤것이라서, 이렇게 짜는 것도 고려해보겠습니다.